Feature/1116 add new check external dependency tasks

Ebaneck opened the pull request:
**Component**:

Buildchain

**Context**: 

Issue: https://github.com/scality/metalk8s/issues/1116

**Summary**:

Improve the management of external dependencies properly and become gentle with traceback printed to the user

**Acceptance criteria**: 

Generate a clean trackback if a required external dependency is missing.

E.g
`TaskError - taskid:check_for:skopeo
command skopeo not found in $PATH

########################################
check_for:skopeo <stdout>:`



---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #ISSUE_NUMBER

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1116-add-new-check-external-dependency-tasks` into
`development/2.0`
* [w/2.1/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.1` 
* [w/2.2/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal

gdemonet commented on buildchain/buildchain/image.py:
```diff
@@ -82,7 +83,7 @@ def show() -> str:
 
     task = targets.Target(task_dep=['_image_pull', '_image_build']).basic_task
```
Would need a dependency on `hardlink` then, no?

```suggestion
     task = targets.Target(
         task_dep=['check_for:hardlink', '_image_pull', '_image_build']
     ).basic_task
```

charlesprost commented:
@Ebaneck just got this, any clue?

```
$ ./doit.sh 
-- MKDIR        _build
-- MKDIR        _build/root
-- MKDIR        _build/root/examples
.  check_for:docker
-- MKDIR        _build/root/images
-- MKTREE       _build/root/pillar/metalk8s/roles
.  CPTREE       pillar
.  RENDER       _build/root/pillar/top.sls
.  MKTREE       _build/root/salt/metalk8s/addons/monitoring/grafana
.  CPTREE       salt
-- RENDER       _build/root/salt/top.sls
-- RENDER       _build/root/salt/metalk8s/kubernetes/mark-control-plane/deployed.sls
-- MKDIR        _build/packages
-- MKDIR        _build/root/packages
-- MKDIR        _build/root/packages/metalk8s-scality-el7
-- MKDIR        _build/root/packages/metalk8s-scality-el7/x86_64
.  GENERATE     _build/root/product.txt
.  RENDER       _build/root/bootstrap.sh
-- GENERATE     _build/root/examples/new-node.yaml
.  check_for:skopeo
-- PULL IMG     coredns:1.3.1
-- PULL IMG     etcd:3.2.18
-- PULL IMG     grafana:5.2.4
-- PULL IMG     kube-apiserver:1.11.7
-- PULL IMG     kube-controller-manager:1.11.7
-- PULL IMG     kube-proxy:1.11.7
-- PULL IMG     kube-rbac-proxy:v0.3.1
-- PULL IMG     kube-scheduler:1.11.7
-- PULL IMG     kube-state-metrics:v1.3.1
-- PULL IMG     nginx:1.15.8
-- PULL IMG     node-exporter:v0.17.0
-- PULL IMG     prometheus:v2.4.3
-- PULL IMG     prometheus-config-reloader:v0.23.2
.  PULL IMG     addon-resizer:1.8.3
TaskError - taskid:_image_pull:addon-resizer
CmdAction Error creating command string
Traceback (most recent call last):
  File "/home/nootal/projects/metalk8s/buildchain/.venv/lib/python3.6/site-packages/doit/action.py", line 196, in execute
    action = self.expand_action()
  File "/home/nootal/projects/metalk8s/buildchain/.venv/lib/python3.6/site-packages/doit/action.py", line 277, in expand_action
    msg % (self.task.name, element, type(element)))
doit.exceptions.InvalidTask: _image_pull:addon-resizer. CmdAction element must be a str or Path from pathlib. Got 'None' (<class 'NoneType'>)

########################################
_image_pull:addon-resizer <stdout>:
```

gdemonet commented on buildchain/buildchain/deps.py:
```diff
@@ -0,0 +1,36 @@
+        If the user already specified a path for it, check for the existence
+        of the binary at the given location. Otherwise, look it up from PATH.
+
+        If the binary is found, the `CMD` dict is updated, otherwise it fails.
```
I'm quite lost. Let me detail the few code paths I understand of your code:

_Case 1:_ `SOME_CMD_BIN` is not defined in the environment, so this task will search in `$PATH` and override `CMD['some-cmd']`

_Case 2_: `SOME_CMD_BIN` is set to `/some/path/some-cmd`, so this task will search in `/some/path` and set `CMD['some-cmd']` to the same value as it was already

_Case 3_: `SOME_CMD_BIN` is set to `/some/path/alias-cmd` (note the different binary name), so this task will search for `some-cmd` in `/some/path`, and **will fail**.

I believe we need to be more declarative and less magic about finding binaries we rely on.

First, using a mutable dict like this prevents static checkers from knowing whether the relevant command is defined in `config.CMD`. We could have used an `enum.Enum` to fix that aspect.

Second, you are mutating the dict at every execution of `check_for:some-cmd`, redundant though not critical.

Finally, I'd rather let the user responsible of managing his commands: the task could just _check_ if the provided path exists, complain if not, and get on with it. No need to mutate anything, nor to run this check for each task that rely on it.

Ebaneck replied:
@gdemonet I would certainly look into the potential use of enumerators.

In regards to the second potential issue of mutating Dicts each time; I think this is pinned to the underline functioning of **Python Doit**.

**step i.** All tasks are created + graphs are built.
**step ii.** Tasks are executed.

What does this mean for us??
Once tasks are created, the dict is initialized with none until the tasks are ran  For example, the dict will only be updated if  `check_for:docker` had ran before hence the mutation:

`TaskError - taskid:check_for:skopeo
command skopeo {'docker': '/usr/bin/docker', 'skopeo': None, 'vagrant': None, 'mkisofs': None, 'hardlink': None} not found in $PATH
`

@slaperche-scality has proposed a better way of handling this and we are looking into this.

And yes with this new approach, we potentially will take off mutations each time using `        'actions': [CmdAction(create_cmd_string)],`

slaperche-scality replied:
> Case 2: SOME_CMD_BIN is set to /some/path/some-cmd, so this task will search in /some/path and set CMD['some-cmd'] to the same value as it was already

Yeah, we don't change the value here.
The point is to check that we indeed have a `some-cmd` and that it's executable, … in the given path.

> No need to mutate anything, nor to run this check for each task that rely on it.
You need to run it for every task that rely on it because you don't know WHICH command will be run, and as said before it should be run only once per execution.

> Second, you are mutating the dict at every execution of check_for:some-cmd, redundant though not critical.

For that point, we could probably have a custom `uptodate` that run the task only if the dict entry is `None`: that way it'll only be run once, even if multiple tasks depends on it.
But I think even as it is, the task is only run once (to be confirmed) so it's unnecessary.

gdemonet replied:
> In regards to the second potential issue of mutating Dicts each time; I think this is pinned to the underline functioning of Python Doit.
>
> **step i.** All tasks are created + graphs are built.
> **step ii.** Tasks are executed.

I don't see how this means you **need** mutability. Most importantly, my point is that we don't need to mutate _anything_.

> For that point, we could probably have a custom `uptodate` that run the task only if the dict entry is `None`: that way it'll only be run once, even if multiple tasks depends on it.
> But I think even as it is, the task is only run once (to be confirmed) so it's unnecessary.

If the task is run once only, then let it fail if the path is not found, and that's enough. No need to store it in a dict (either it's in `$PATH`, or it's provided by `os.getenv("XXX_BIN")`).

gdemonet replied:
Also, how would this behaviour of mutating dicts work in a concurrent execution? Is `doit` responsible for passing the shared state to spawned tasks?

slaperche-scality replied:
> If the task is run once only, then let it fail if the path is not found, and that's enough. No need to store it in a dict (either it's in $PATH, or it's provided by os.getenv("XXX_BIN")).

Hum, fair enough.
This will simplify the code!

> Also, how would this behaviour of mutating dicts work in a concurrent execution? Is doit responsible for passing the shared state to spawned tasks?

Good point.

gdemonet reviewed (changes requested):
Good usability improvement, although I think we're missing a situation where the user wants to define an alias for some binaries (rename `vagrant` to `vagrant-X.Y` for instance).

Another point, though I'm not sure it's a use-case we want to support, we can't use the `XXX_BIN` env vars to add options, such as verbosity.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

Ebaneck commented:
> @Ebaneck just got this, any clue?
> 
> ```
> $ ./doit.sh 
> -- MKDIR        _build
> -- MKDIR        _build/root
> -- MKDIR        _build/root/examples
> .  check_for:docker
> -- MKDIR        _build/root/images
> -- MKTREE       _build/root/pillar/metalk8s/roles
> .  CPTREE       pillar
> .  RENDER       _build/root/pillar/top.sls
> .  MKTREE       _build/root/salt/metalk8s/addons/monitoring/grafana
> .  CPTREE       salt
> -- RENDER       _build/root/salt/top.sls
> -- RENDER       _build/root/salt/metalk8s/kubernetes/mark-control-plane/deployed.sls
> -- MKDIR        _build/packages
> -- MKDIR        _build/root/packages
> -- MKDIR        _build/root/packages/metalk8s-scality-el7
> -- MKDIR        _build/root/packages/metalk8s-scality-el7/x86_64
> .  GENERATE     _build/root/product.txt
> .  RENDER       _build/root/bootstrap.sh
> -- GENERATE     _build/root/examples/new-node.yaml
> .  check_for:skopeo
> -- PULL IMG     coredns:1.3.1
> -- PULL IMG     etcd:3.2.18
> -- PULL IMG     grafana:5.2.4
> -- PULL IMG     kube-apiserver:1.11.7
> -- PULL IMG     kube-controller-manager:1.11.7
> -- PULL IMG     kube-proxy:1.11.7
> -- PULL IMG     kube-rbac-proxy:v0.3.1
> -- PULL IMG     kube-scheduler:1.11.7
> -- PULL IMG     kube-state-metrics:v1.3.1
> -- PULL IMG     nginx:1.15.8
> -- PULL IMG     node-exporter:v0.17.0
> -- PULL IMG     prometheus:v2.4.3
> -- PULL IMG     prometheus-config-reloader:v0.23.2
> .  PULL IMG     addon-resizer:1.8.3
> TaskError - taskid:_image_pull:addon-resizer
> CmdAction Error creating command string
> Traceback (most recent call last):
>   File "/home/nootal/projects/metalk8s/buildchain/.venv/lib/python3.6/site-packages/doit/action.py", line 196, in execute
>     action = self.expand_action()
>   File "/home/nootal/projects/metalk8s/buildchain/.venv/lib/python3.6/site-packages/doit/action.py", line 277, in expand_action
>     msg % (self.task.name, element, type(element)))
> doit.exceptions.InvalidTask: _image_pull:addon-resizer. CmdAction element must be a str or Path from pathlib. Got 'None' (<class 'NoneType'>)
> 
> ########################################
> _image_pull:addon-resizer <stdout>:
> ```

@nootal This is a work in progress... We are currently refactoring some of the logic + this is marked a draft PR.. Will ping you when it is fully ready for review

slaperche-scality commented:
> Good usability improvement, although I think we're missing a situation where the user wants to define an alias for some binaries (rename `vagrant` to `vagrant-X.Y` for instance).

If someone has an alias, it can uses the `XXX_BIN`, no?
That's the point of those actually, to handle case when we are not in the default setup (binary with default name in the `PATH`).

> Another point, though I'm not sure it's a use-case we want to support, we can't use the `XXX_BIN` env vars to add options, such as verbosity.

It more or less supported for `vagrant` (with `VAGRANT_UP_ARGS`), but not sure we want to go that way.
We could also leverage the `verbose` options of `doit` and patch the commands ourself when it's set.
Or in some other way.

Anyway, this is not related to this PR (but could be tracked in a dedicated issue :slightly_smiling_face: )

gdemonet commented:
> > Good usability improvement, although I think we're missing a situation where the user wants to define an alias for some binaries (rename `vagrant` to `vagrant-X.Y` for instance).
> 
> If someone has an alias, it can uses the `XXX_BIN`, no?
> That's the point of those actually, to handle case when we are not in the default setup (binary with default name in the `PATH`).

No, that's my point actually (see https://github.com/scality/metalk8s/pull/1233#discussion_r294186349):

[...] 3. `SOME_CMD_BIN` is set to `/some/path/alias-cmd` (note the different binary name), so this task will search for `some-cmd` in `/some/path`, and **will fail**.


> > Another point, though I'm not sure it's a use-case we want to support, we can't use the `XXX_BIN` env vars to add options, such as verbosity.
> 
> It more or less supported for `vagrant` (with `VAGRANT_UP_ARGS`), but not sure we want to go that way.
> We could also leverage the `verbose` options of `doit` and patch the commands ourself when it's set.
> Or in some other way.
> 
> Anyway, this is not related to this PR (but could be tracked in a dedicated issue  )

Agreed, let's not focus on this right now.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @nicolast

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @cmonfort

* @alxf

* @alexandre-allard-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @monpote

* @jbertran

* @gdemonet

* @slaperche-scality

* @thomasdanan

* @ramanduh





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

slaperche-scality commented:
> /some/path/alias-cmd



> No, that's my point actually (see [#1233 (comment)](https://github.com/scality/metalk8s/pull/1233#discussion_r294186349)):
> 
> [...] 3. `SOME_CMD_BIN` is set to `/some/path/alias-cmd` (note the different binary name), so this task will search for `some-cmd` in `/some/path`, and **will fail**.

If someone tells use `look for /some/path/alias-cmd`, of course we aren't going to find `some-cmd` (because they explicitly told us to look for `alias-cmd`).

slaperche-scality commented on buildchain/buildchain/config.py:
```diff
@@ -47,11 +47,18 @@
 # Name of the command (if in the PATH) or path to the binary.
 
-DOCKER   : str = os.getenv('DOCKER_BIN',   'docker')
+# DOCKER   : str = os.getenv('DOCKER_BIN',   'docker')
```
Following @gdemonet recommendation, we will give up the `dict`-based approach and move onto `enum`.

This will simplify the code (we no longer need to use `CmdAction`).

This should become:

```python
# External commands {{{

# Name of the external commands (if in the PATH) or path to the binary.

class ExtCommand(enum.Enum):
    """External commands used by the build chain."""

    GIT      = os.getenv('GIT_BIN',      'git')
    HARDLINK = os.getenv('HARDLINK_BIN', 'hardlink')
    MKISOFS  = os.getenv('MKISOFS_BIN',  'mkisofs')
    SKOPEO   = os.getenv('SKOPEO_BIN',   'skopeo')
    VAGRANT  = os.getenv('VAGRANT_BIN',  'vagrant')

# }}}
```

Ebaneck replied:
I think that the GIT_BIN should not be included here. no??

slaperche-scality commented on buildchain/buildchain/deps.py:
```diff
@@ -0,0 +1,36 @@
+from buildchain import config
+
+
+def task_check_for():
```
Following @gdemonet  recommendations, this would become

```python
def task_check_for():
    """Check the existence of the external commands."""
    def get_command_location(
        command_name: str, binary_path: Optional[str]
    ) -> Optional[TaskError]:
        """Check the existence of the given command.

        If the user already specified a path for it, check for the existence
        of the binary at the given location. Otherwise, look it up from PATH.

        If the binary is not found, the task fails.
        """
        if binary_path:
            path = Path(binary_path)
            cmd_path = shutil.which(path.name, path=path.parent)
        else:
            cmd_path = shutil.which(command_name)
        if cmd_path is None:
            return TaskError(msg='command {} not found in {}'.format(
                command_name, Path(binary_path).parent
                if binary_path else '$PATH'
            ))
        return None

    for ext_cmd in config.ExtCommand:
        cmd_name = ext_cmd.name.lower()
        cmd_path = ext_cmd.value if ext_cmd.value != cmd_name else None

        def show(name) -> str:
            return '{cmd: <{width}} {name}'.format(
                cmd='CHECK CMD', width=constants.CMD_WIDTH, name=name
            )

        yield {
            'name': cmd_name,
            'title': lambda _, name=cmd_name: show(name),
            'doc': 'Check the presence of the {} command.'.format(cmd_name),
            'actions': [(get_command_location, [cmd_name, cmd_path], {})],
            'file_dep': [],
            'task_dep': [],
            'uptodate': [False],
        }
```

I added a `title` field for a nice output + we  no longer update a `dict`.

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -79,10 +79,10 @@ def show() -> str:
         )
-
-    task = targets.Target(task_dep=['_image_pull', '_image_build']).basic_task
+    task = targets.Target(task_dep=['check_for:hardlink',
```
This can be wrapped as:

```
    task = targets.Target(
        task_dep=['_image_pull', '_image_build', 'check_for:hardlink']
    ).basic_task
```

which is a bit more readable.

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -79,10 +79,10 @@ def show() -> str:
+                                    '_image_pull', '_image_build']).basic_task
     task['title']   = lambda _: show()
-    task['actions'] = [[config.HARDLINK, '-c', constants.ISO_IMAGE_ROOT]]
+    task['actions'] = [[config.CMD['hardlink'], '-c', constants.ISO_IMAGE_ROOT]]
```
Using the enum approach, this would become `config.ExtCommand.HARDLINK.value`

slaperche-scality commented on buildchain/buildchain/deps.py:
```diff
@@ -0,0 +1,36 @@
+from doit.exceptions import TaskError
```
Import should be split in 3 groups:
- standard library
- third-party library
- local module

e.g.:

```
from pathlib import Path
from typing import Optional
import shutil

from doit.exceptions import TaskError

from buildchain import config
from buildchain import constants
from buildchain import utils
```

slaperche-scality replied:
You need a docstring to describe what your module is doing, e.g.: `"""Tasks to check the dependencies on external commands."""` as the first line of the file

slaperche-scality commented on buildchain/buildchain/iso.py:
```diff
@@ -160,7 +160,7 @@ def action(targets: Sequence[str]) -> None:
     """Create the ISO from the files in ISO_ROOT."""
     mkisofs = [
-        config.MKISOFS, '-output',  ISO_FILE,
+        config.CMD['mkisofs'], '-output',  ISO_FILE,
```
`config.ExtCommand.MKISOFS.value`

slaperche-scality commented on buildchain/buildchain/targets/local_image.py:
```diff
@@ -169,6 +170,7 @@ def task(self) -> types.TaskDict:
             'title': lambda _: self.show('IMG BUILD'),
             'doc': 'Build {} container image.'.format(self.name),
             'actions': self._build_actions(),
+            'task_dep': ['check_for:skopeo', 'check_for:docker'],
```
No need for `check_for:docker` + you shouldn't assign `task_dep` directly: if the user specified its own `task_dep` you'll overwrite them.

I recommend to append to `task_dep`, in the `__init__`:

```diff
@@ -107,6 +107,7 @@ class LocalImage(image.ContainerImage):
         self._save = save_on_disk
         self._build_args = build_args or {}
         kwargs.setdefault('file_dep', []).append(self.dockerfile)
+        kwargs.setdefault('task_dep', []).append('check_for:skopeo')
         super().__init__(
             name=name, version=version, destination=destination, **kwargs
         )
```

gdemonet commented:
> > /some/path/alias-cmd
> 
> > No, that's my point actually (see [#1233 (comment)](https://github.com/scality/metalk8s/pull/1233#discussion_r294186349)):
> > [...] 3. `SOME_CMD_BIN` is set to `/some/path/alias-cmd` (note the different binary name), so this task will search for `some-cmd` in `/some/path`, and **will fail**.
> 
> If someone tells use `look for /some/path/alias-cmd`, of course we aren't going to find `some-cmd` (because they explicitly told us to look for `alias-cmd`).

Yes, the code needs to be fixed.

slaperche-scality commented on buildchain/buildchain/targets/local_image.py:
```diff
@@ -177,6 +179,21 @@ def task(self) -> types.TaskDict:
 
+    def _build_skopeo_cmd(self) -> List[str]:
+        """Return the command line to execute skopeo cmd."""
+        cmd = [
```
```python
            cmd = [
                config.ExtCommand.SKOPEO.value,
                '--override-os', 'linux', '--insecure-policy',
                'copy', '--format', 'v2s2', '--dest-compress',
            ]
```

slaperche-scality commented on buildchain/buildchain/targets/remote_image.py:
```diff
@@ -58,6 +60,7 @@ def __init__(
             **kwargs
         )
         self._targets = [self.filepath]
+        self.task_dep.append('check_for:skopeo')
```
I would use the same pattern as for `local_image`, e.g:

```diff
+++ b/buildchain/buildchain/targets/remote_image.py
@@ -52,6 +52,7 @@ class RemoteImage(image.ContainerImage):
         self._digest = digest
         self._remote_name = remote_name or name
         self._use_tar = save_as_tar
+        kwargs.setdefault('task_dep', []).append('check_for:skopeo')
         super().__init__(
             name=name, version=version,
             destination=destination,
```

slaperche-scality commented on buildchain/buildchain/targets/remote_image.py:
```diff
@@ -92,19 +95,23 @@ def task(self) -> types.TaskDict:
         """Return the command line to execute skopeo copy."""
         cmd = [
-            config.SKOPEO, '--override-os', 'linux', '--insecure-policy',
+            config.CMD['skopeo'], '--override-os', 'linux', '--insecure-policy',
```
```
        cmd = [
            config.ExtCommand.SKOPEO.value,
            '--override-os', 'linux', '--insecure-policy',
            'copy', '--format', 'v2s2'
        ]
```

slaperche-scality commented on buildchain/buildchain/deps.py:
```diff
@@ -0,0 +1,36 @@
+            'file_dep': [],
+            'task_dep': [],
+            'uptodate': [False],
+        }
```
You need to add `__all__ = utils.export_only_tasks(__name__)` at the end of the file, to only export the tasks (it's also done for the others, see `iso.py` for instance).

slaperche-scality commented on buildchain/buildchain/vagrant.py:
```diff
@@ -38,55 +39,62 @@ def mkdir_dot_vagrant() -> None:
 
     return {
-        'actions': [doit.tools.LongRunning(' '.join(cmd))],
+        'actions': [doit.action.CmdAction(_build_vagrant_cmd, shell=True)],
```
Since we no longer use a dict, you can keep the `LongRunning` as it was before.

slaperche-scality commented on buildchain/buildchain/vagrant.py:
```diff
@@ -38,55 +39,62 @@ def mkdir_dot_vagrant() -> None:
-    cmd = [config.VAGRANT, 'up']
+    
+def _build_vagrant_cmd(self) -> List[str]:
+    cmd = [config.CMD['vagrant'], 'up']
```
`s/config.CMD['vagrant']/config.ExtCommand.VAGRANT.value/`, same for the other occurrence in the file.

slaperche-scality reviewed (changes requested):
We will follow @gdemonet good recommendation on getting rid of the `dict`: this will simplify our code (no longer need to use `CmdAction`).

Also, I think it would be nice if you PR was split in 3 commits:
- the first to delete `DOCKER   : str = os.getenv('DOCKER_BIN',   'docker')` because it's no longer used since Jordi's PR (should be mentionned in the commit message)
- the second one to introduce the enum and `deps.py`
- the third one to replace the `config.XXX` by `config.ExtCommand.XXX.value` and adding the `task_dep`

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @nicolast

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @cmonfort

* @alxf

* @alexandre-allard-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @monpote

* @jbertran

* @gdemonet

* @slaperche-scality

* @thomasdanan

* @ramanduh





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

* @slaperche-scality

Ebaneck committed:
build: remove DOCKER_BIN

Since the introduction of PR #939, we use the Docker API instead of the binary
directly, thus it's no longer used and can be safely remove.

Ebaneck committed:
build: add a task to check for external binaries

This commit introduces a new set of tasks to check for the presence
of some required binaries.
If a required binary is missing then the task fails, build is stopped
and a clear error message is printed instead of a huge stacktrace.

Ebaneck committed:
build: add task dep on binaries checking

For every task that relies on an external command, we add a dependence
to the task that checks the existence of the required binary.

bert-e commented:
# History mismatch



Merge commit #27437abfa01cffb662f8e7831edddb511006027c on the integration branch
`w/2.1/feature/1116-add-new-check-external-dependency-tasks` is merging a branch which is neither the current
branch `feature/1116-add-new-check-external-dependency-tasks` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1116-add-new-check-external-dependency-tasks` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1116-add-new-check-external-dependency-tasks` into
`development/2.0`
* [w/2.1/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.1` 
* [w/2.2/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

* @gdemonet

slaperche-scality commented on buildchain/buildchain/config.py:
```diff
@@ -47,7 +47,6 @@
 
 # Name of the command (if in the PATH) or path to the binary.
 
-DOCKER   : str = os.getenv('DOCKER_BIN',   'docker')
```
Comments about the commit message:

The PR that introduced the docker API is #939, not #1191 (which is a work in progress).

`donot` => either `don't` or `do not`.

e.g.:

```
build: remove DOCKER_BIN

Since the introduction #939, we use the Docker API instead of the binary
directly, thus it's no longer used and can be safely removed.
```

slaperche-scality commented on buildchain/buildchain/deps.py:
```diff
@@ -1,3 +1,7 @@
+
```
No need to skip a line here, the docstring can be the first line of the file (IIRC, that's what we do in the other files).

slaperche-scality commented on buildchain/buildchain/config.py:
```diff
@@ -9,6 +9,7 @@
 
 import os
 import shlex
+import enum
```
Commit message could be more explicit, like:

```
build: add a test to check for external binaries

This commit introduces a new set of tasks to check for the presence of some required binaries.
If a required binary is missing, the task will fail and stop the build with a clear error message (instead of a big stacktrace).
```

Otherwise, the code LGTM.

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -80,9 +80,13 @@ def show() -> str:
         )
 
-    task = targets.Target(task_dep=['_image_pull', '_image_build']).basic_task
+    task = targets.Target(
```
minor: code layout could be improved a little bit

Align closing parenthesis:
```
    task = targets.Target(
        task_dep=['check_for:hardlink', '_image_pull', '_image_build']
    ).basic_task
```

Try to keep the command on a single line:

```
task['actions'] = [
    [config.ExtCommand.HARDLINK.value, '-c', constants.ISO_IMAGE_ROOT]
]
```

Ebaneck replied:
Great, I learnt we have a vertical ruler of 79 enforced. This is the only reason I reformated this. I am going to keep the single line

slaperche-scality commented on buildchain/buildchain/constants.py:
```diff
@@ -91,8 +91,8 @@ def git_ref() -> Optional[str]:
         ref : bytes = subprocess.check_output([
-            config.GIT, 'describe', '--always', '--long', '--tags',
-                '--dirty',
+            config.ExtCommand.GIT.value, 'describe', '--always', '--long',
```
about commit message: the reference to the issue is usually put at the end of the commit, not in the title.
Also, we usually prefix the title with the component/module we are working on (in that case `build` or `buildchain`).

e.g.:

```
build: add task dep on binaries checking

For every task that rely on an external command, we add a dependence to the task that check the existence of the required binary.
```

slaperche-scality reviewed (changes requested):
Code is pretty good (only have a few minor remarks), most comments are about commit messages.

bert-e commented:
# History mismatch



Merge commit #c093861ca6a821ce4a1d9e1e824b261883d6b327 on the integration branch
`w/2.1/feature/1116-add-new-check-external-dependency-tasks` is merging a branch which is neither the current
branch `feature/1116-add-new-check-external-dependency-tasks` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1116-add-new-check-external-dependency-tasks` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
@bert-e reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**

bert-e commented:
# History mismatch



Merge commit #c093861ca6a821ce4a1d9e1e824b261883d6b327 on the integration branch
`w/2.1/feature/1116-add-new-check-external-dependency-tasks` is merging a branch which is neither the current
branch `feature/1116-add-new-check-external-dependency-tasks` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1116-add-new-check-external-dependency-tasks` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
@bert-e force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1116-add-new-check-external-dependency-tasks` into
`development/2.0`
* [w/2.1/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.1` 
* [w/2.2/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

* @gdemonet

bert-e commented:
# History mismatch



Merge commit #711f0e2edf7004fb1a9e94963c6b873874f1df14 on the integration branch
`w/2.1/feature/1116-add-new-check-external-dependency-tasks` is merging a branch which is neither the current
branch `feature/1116-add-new-check-external-dependency-tasks` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1116-add-new-check-external-dependency-tasks` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1116-add-new-check-external-dependency-tasks` into
`development/2.0`
* [w/2.1/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.1` 
* [w/2.2/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

gdemonet commented on buildchain/buildchain/image.py:
```diff
@@ -80,9 +80,13 @@ def show() -> str:
-    task = targets.Target(task_dep=['_image_pull', '_image_build']).basic_task
+    task = targets.Target(
+        task_dep=['check_for:hardlink', '_image_pull', '_image_build']
+        ).basic_task
```
This one should be unindented:

```suggestion
    ).basic_task
```

gdemonet reviewed (dismissed):
Great work, thanks :) Will be much friendlier for newcomers with this addition!

(just a minor linting comment)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #3fa5112a08249d8dab3fe0253ffa35ff9b06477d on the integration branch
`w/2.1/feature/1116-add-new-check-external-dependency-tasks` is merging a branch which is neither the current
branch `feature/1116-add-new-check-external-dependency-tasks` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1116-add-new-check-external-dependency-tasks` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1116-add-new-check-external-dependency-tasks` into
`development/2.0`
* [w/2.1/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.1` 
* [w/2.2/feature/1116-add-new-check-external-dependency-tasks](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1116-add-new-check-external-dependency-tasks) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

Ebaneck commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
